### ISSUE TITLE
Consider making database functions more first-class in the model

### ISSUE THREAD

**Issue by divega** [aspnet/EntityFramework#9393]
(This is based on the discussion in https://github.com/aspnet/EntityFramework/issues/9368)

We could:
- Allow functions to be declared in the model, likely based on the work already implemented by @pmiddleton but the ability to specify an accompanying database implementation (e.g. function body) for them
- Add first class ways to reference these declared functions in parts of the model, as part of computed properties on an entity as a function of the PK of the entity or of a path of properties starting with the entity. E.g. a customer could have a computed property that is computed as the SUM() of the values of all his/her orders.
- Add the capability in the model differ to compare function implementations and references to functions
- Add the capability in DDL generation to created, drop, and alter functions, as well as references to functions, so that both migrations and `EnsureCreated()` can start generating those.

**Comment by roji** [comment]
If you go ahead with this, it's very important to have other databases in mind, as function support does vary between different systems ([here is the PostgreSQL `CREATE FUNCTION` documentation](https://www.postgresql.org/docs/current/static/sql-createfunction.html)).

**Comment by divega** [comment]
@roji Indeed, we would need to triangulate something that makes sense across various databases and make it extensible for providers. 

BTW, do you have any feedback on how useful this could be? Currently you can do most things by resorting to raw SQL in migrations, but this would be more abstract and declarative support.

**Comment by roji** [comment]
@divega I'm not sure I'm the one to ask, I'm just the driver guy :) I personally avoid putting code in my databases, because I like to code in safe, static languages such as C# - unless there's a real performance argument for saving costly network roundtrips by performing logic in the database (or some rare situation where it makes sense to share logic between multiple clients at the database side rather than at the client side). As an O/RM, EF Core encourages writing logic in C# rather than in SQL (or any other database-specific language), so this feature goes in a pretty different direction.

More to the point though, here are some negative points:

* One of the strongest points of EF Core is creating (and evolving!) a database schema as a function of changes in the CLR entity - you change a C# class and your table changes automatically. With functions there's no CLR entity: the function can only change when the developer manually modifies their definition in the model, IMHO this limits the usefulness of this feature. To make the point clear, if you had considered a feature which automatically creates a database function based on a C# expression tree - and therefore evolves the former when the latter changes - you'd have something truly valuable, but that's a whole other project (which is likely impossible/impractical for various reasons).
* Another point about table migrations is that they can evolve *part* of a table schema (e.g. add a column). Functions, on the other hand, are a simple thing which likely wouldn't be altered, but rather dropped and recreated with a new definition. That seems quite limited - as far as creating and evolving the function is concerned, there's simply not much value.

However:

* Regardless of what I think about user-defined database functions, many people use them and supporting is important.
* The ability to statically invoke a function from a LINQ query is a definite big win.

So I'd say it's definitely a good feature for those people who do like/need to put logic in their database (but perhaps not a very high-priority one).

**Comment by saadhaffar025** [comment]
Hi,
In your context constructor, you can write the following:
public MyContext()
            : base("MyDatabaseConnectionString")
        {
            this.Database.ExecuteSqlCommand("some sql here"); // Or make this code inside a C# method if you have a lot of functions and stored procedures to create.
        }

**Comment by dbrownems** [comment]
Apart from the enabling code-first, the EF Core lacks the ability to join TVFs.  In EF6 an Entity attribute can be passed to a TVF

```
from c in db.Customers
from cd in db.CustomerDemographics(c.CustomerId)
select ...
```
would generate SQL like:

```
select ...
from customer c
cross apply dbo.customer_demographics(c.customer_id) cd

```

But in EF Core, the best you can do is call `FromSqlInterpolated` which doesn't enable this scenario.
